relayer/relays/beacon/header/header.go

```diff
@@ -0,0 +1,262 @@
+package header
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"time"
+
+	"github.com/ethereum/go-ethereum/common"
+	log "github.com/sirupsen/logrus"
+	"github.com/snowfork/go-substrate-rpc-client/v4/types"
+	"github.com/snowfork/snowbridge/relayer/relays/beacon/cache"
+	"github.com/snowfork/snowbridge/relayer/relays/beacon/header/syncer"
+	"github.com/snowfork/snowbridge/relayer/relays/beacon/header/syncer/scale"
+	"github.com/snowfork/snowbridge/relayer/relays/beacon/writer"
+)
+
+var ErrFinalizedHeaderUnchanged = errors.New("finalized header unchanged")
+
+type Header struct {
+	cache  *cache.BeaconCache
+	writer *writer.ParachainWriter
+	syncer *syncer.Syncer
+}
+
+func New(writer *writer.ParachainWriter, beaconEndpoint string, slotsInEpoch uint64, epochsPerSyncCommitteePeriod uint64) Header {
+	return Header{
+		cache:  cache.New(),
+		writer: writer,
+		syncer: syncer.New(beaconEndpoint, slotsInEpoch, epochsPerSyncCommitteePeriod),
+	}
+}
+
+func (h *Header) Sync(ctx context.Context) (<-chan uint64, <-chan uint64, error) {
+	latestSyncedPeriod, err := h.writer.GetLastSyncedSyncCommitteePeriod()
+	if err != nil {
+		return nil, nil, fmt.Errorf("fetch last sync commitee: %w", err)
+	}
+
+	h.cache.SetLastSyncedSyncCommitteePeriod(latestSyncedPeriod)
+
+	log.WithField("period", latestSyncedPeriod).Info("set cache: last beacon synced sync committee period")
+
+	periodsToSync, err := h.syncer.GetSyncPeriodsToFetch(latestSyncedPeriod)
+	if err != nil {
+		return nil, nil, fmt.Errorf("check sync committee periods to be fetched: %w", err)
+	}
+
+	log.WithFields(log.Fields{
+		"periods": periodsToSync,
+	}).Info("sync committee periods to be synced")
+
+	for _, period := range periodsToSync {
+		err := h.SyncCommitteePeriodUpdate(ctx, period)
+		if err != nil {
+			return nil, nil, err
+		}
+	}
+
+	lastFinalizedHeader, err := h.writer.GetLastStoredFinalizedHeader()
+	if err != nil {
+		return nil, nil, fmt.Errorf("fetch last finalized header: %w", err)
+	}
+
+	lastFinalizedSlot, err := h.writer.GetLastStoredFinalizedHeaderSlot()
+	if err != nil {
+		return nil, nil, fmt.Errorf("fetch last finalized header slot: %w", err)
+	}
+
+	h.cache.FinalizedHeaders = append(h.cache.FinalizedHeaders, lastFinalizedHeader)
+
+	log.WithFields(log.Fields{
+		"hash": lastFinalizedHeader,
+		"slot": lastFinalizedSlot,
+	}).Info("set cache: last finalized header")
+
+	log.Info("starting to sync finalized headers")
+
+	ticker := time.NewTicker(time.Second * 20)
+	done := make(chan bool)
+
+	basicChannel := make(chan uint64)
+	incentivizedChannel := make(chan uint64)
+
+	go func() {
+		for {
+			lastBlockNumber, err := h.SyncHeaders(ctx)
+			switch {
+			case errors.Is(err, ErrFinalizedHeaderUnchanged):
+				log.WithField("finalized_header", h.cache.LastFinalizedHeader()).Info("finalized header unchanged")
+			case err != nil:
```

Comment:
If we hit this error case, the goroutine will silently die, and then no more headers will be sync'ed until the process is restarted.

This has been a common issue in our relayer during its development, so we've moved over to spawning all goroutines using [errgroup](https://pkg.go.dev/golang.org/x/sync/errgroup). It allows the relayer to detect a failed goroutine and self-terminate itself (roughly equivalent to a  fail-fast error handling approach).

Can see an example of this approach in our old ethereum relay:
1. Spawn goroutine using errgroup: https://github.com/Snowfork/snowbridge/blob/f719545fe8d8cf9b5e0665b690c73077ac49258c/relayer/relays/ethereum/ethereum-listener.go#L125
2. If the goroutine is meant to run for ever, make sure it can self-terminate if the associated context.Context is done. (This is needed for orderly termination of all goroutines if one of them fails, or the process is signalled with SIGINT or something like that): https://github.com/Snowfork/snowbridge/blob/f719545fe8d8cf9b5e0665b690c73077ac49258c/relayer/relays/ethereum/ethereum-listener.go#L150
3. The process waits until all goroutines complete or one of them fails: https://github.com/Snowfork/snowbridge/blob/f719545fe8d8cf9b5e0665b690c73077ac49258c/relayer/cmd/run/ethereum/command.go#L97

Reply:
Good point, thanks for catching this! https://github.com/Snowfork/snowbridge/pull/671/commits/80b7345618449fb8643e4b8b90ed12486efbe073